presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneTableScanColumns.java

```diff
@@ -13,34 +13,109 @@
  */
 package io.prestosql.sql.planner.iterative.rule;
 
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.Session;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.TableHandle;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.ProjectionApplicationResult;
+import io.prestosql.spi.expression.ConnectorExpression;
+import io.prestosql.spi.expression.Variable;
+import io.prestosql.sql.planner.ConnectorExpressionTranslator;
 import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.TypeProvider;
 import io.prestosql.sql.planner.plan.PlanNode;
 import io.prestosql.sql.planner.plan.TableScanNode;
 
+import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
-import static com.google.common.collect.Maps.filterKeys;
+import static com.google.common.collect.ImmutableList.toImmutableList;
 import static io.prestosql.sql.planner.plan.Patterns.tableScan;
-import static io.prestosql.util.MoreLists.filteredCopy;
+import static java.util.Objects.requireNonNull;
 
 public class PruneTableScanColumns
         extends ProjectOffPushDownRule<TableScanNode>
 {
-    public PruneTableScanColumns()
+    private final Metadata metadata;
+    private final TypeAnalyzer typeAnalyzer;
+
+    public PruneTableScanColumns(Metadata metadata, TypeAnalyzer typeAnalyzer)
```

Comment:
what's the relation between this rule and PushProjectionIntoTableScan?

Reply:
That's a good question. PushProjectionIntoTableScan is more general, as it can push down arbitrary sub-expressions from the projection. We'll probably want to remove this one, but I didn't want to conflate issues in this PR.